client/js/lounge.js

```diff
@@ -410,77 +412,235 @@ $(function() {
+		}, defaults);
+
+		this.serverOptions = {};
+
+		this.on("change", function(event, key, value) {
+			self.trigger("change:" + key, [value, key]);
+		});
+
+		this.on("sync", function(event, key, value) {
+			self.trigger("sync:" + key, [value, key]);
+		});
+	}
+
+	Settings.prototype = {
+		get: function(key, def) {
+			if (key in this.serverOptions) {
+				return this.serverOptions[key];
 			}
-			settings.find("#user-specified-css-input").val(options[i]);
-			continue;
-		}
-		if (options[i]) {
-			settings.find("input[name=" + i + "]").prop("checked", true);
+			else if (key in this.options) {
+				return this.options[key];
+			}
+			else {
+				return def;
+			}
+		},
+
+		set: function(key, value) {
+			this.options[key] = value;
+
+			if (key in this.serverOptions) {
+				this.serverOptions[key] = this.options[key] = value;
+			}
+			else {
+				this.options[key] = value;
+			}
+
+			this.trigger("change", [key, value]);
+		},
+
+		sync: function(key, status) {
+			if (status) {
+				this.serverOptions[key] = this.options[key];
+			}
+			else {
+				delete this.serverOptions[key];
+			}
+
+			this.trigger("sync", [key, !!status]);
+		},
+
+		isSync: function(key) {
+			return key in this.serverOptions;
+		},
+
+		on: function(event, callback) {
+			this.element.on("setting:" + event, callback);
+		},
+
+		trigger: function(event, data) {
+			this.element.trigger("setting:" + event, data);
+		},
+
+		mergeFromServer: function(data) {
+			for (var key in this.serverOptions) {
+				this.sync(key, false);
+			}
+
+			for (key in data) {
+				this.set(key, data[key]);
+				this.sync(key, true);
+			}
+		},
+
+		loadFromLocalStorage: function() {
+			var options = JSON.parse(window.localStorage.getItem("settings"));
+
+			for (var k in options) {
+				this.set(k, options[k]);
+			}
+		},
+
+		saveToLocalStorage: function() {
+			window.localStorage.setItem("settings", JSON.stringify(this.options));
 		}
-	}
+	};
 
-	settings.on("change", "input, textarea", function() {
-		var self = $(this);
-		var name = self.attr("name");
+	var settings = new Settings($("#settings"));
 
-		if (self.attr("type") === "checkbox") {
-			options[name] = self.prop("checked");
-		} else {
-			options[name] = self.val();
+	// Bind the settings page inputs
+	(function() {
+		function bindCheckbox(el, name) {
+			el.prop("checked", settings.get(name));
+			el.change(function() {
+				settings.set(name, el.prop("checked"));
+			});
+			settings.on("change:" + name, function(event, value) {
+				el.prop("checked", !!value);
+			});
+		}
+
+		function bindOther(el, name) {
+			el.prop("value", settings.get(name));
+			el.change(function() {
+				settings.set(name, el.prop("value"));
+			});
+			settings.on("change:" + name, function(event, value) {
+				el.prop("value", value);
+			});
 		}
 
-		window.localStorage.setItem("settings", JSON.stringify(options));
+		settings.element.find("input[name^='setting:'], textarea[name^='setting:']").each(function() {
```

Comment:
I could but since this is a two-way sync, I'd need to keep querying and checking the type of element for every setting change. I thought it was clearer that way, especially since we don't have that many boxes to bind.